File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.java
class Solution {
    public int countCharacters(String[] words, String chars) {
        int sum = 0;
        boolean isValid = true;
        
        int[] letters = new int[26];
        for(int i=0; i < chars.length(); i++)
        letters[chars.charAt(i)-'a']++;
        int[] orig = letters.clone();

        for(int i =0; i<words.length; i++){
            for(int j = 0; j<words[i].length(); j++){
                if(letters[words[i].charAt(j)-'a']==0){
                    isValid=false;
                    break;
                }
                letters[words[i].charAt(j)-'a']--;
            }
            if(isValid) sum+=words[i].length();
            isValid=true;
            letters=orig.clone();
        }
        return sum;
    }
}

File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        count1 = {}

        # Frequency of chars
        for i in range(len(chars)):
            count1[chars[i]] = count1.get(chars[i], 0) + 1

        answer = 0

        # Check every word
        for j in words:
            count2 = {}

            # Frequency of current word
            for k in range(len(j)):
                count2[j[k]] = count2.get(j[k], 0) + 1

            canForm = True

            # Compare frequencies
            for ch in count2:
                if count2[ch] > count1.get(ch, 0):
                    canForm = False
                    break

            if canForm:
                answer += len(j)

        return answer